### PR TITLE
add support for NINJA_STATUS variable

### PR DESCRIPTION
**Pull request by E5ten**

### CODE REVIEW

File: build.c [hunk 1]
```diff
@@ -199,6 +202,46 @@ buildadd(struct node *n)
 	e->flags &= ~FLAG_CYCLE;
 }
 
+static void
+statusfmt(char *fmt)
+{
+	struct timespec ctime;
+
+	for (; *fmt; ++fmt)
```
Comment (michaelforney) [inline build.c]: The brace style I prefer is to use braces if the body is composed of multiple lines (even if they are just a single statement), or if any of the other branches of the if-else chain needed braces.

File: build.c [hunk 2]
```diff
@@ -199,6 +202,46 @@ buildadd(struct node *n)
 	e->flags &= ~FLAG_CYCLE;
 }
 
+static void
+statusfmt(char *fmt)
```
Comment (michaelforney) [inline build.c]: Let's call this `printstatus`.

File: build.c [hunk 3]
```diff
@@ -238,7 +281,8 @@ jobstart(struct job *j, struct edge *e)
 		description = buildopts.verbose ? NULL : edgevar(e, "description", true);
 		if (!description || description->n == 0)
 			description = j->cmd;
-		printf("[%zu/%zu] %s\n", nstarted, ntotal, description->s);
+		statusfmt(ninja_status);
+		printf("%s\n", description->s);
```
Comment (michaelforney) [inline build.c]: `puts(description->s)` is a bit more direct, I think I prefer that.

File: build.c [hunk 4]
```diff
@@ -29,8 +30,10 @@ struct job {
 
 struct buildoptions buildopts = {.maxfail = 1};
 static struct edge *work;
-static size_t nstarted, ntotal;
+static size_t nstarted, nfinished, ntotal;
 static bool consoleused;
+static char *ninja_status;
+static struct timespec stime;
```
Comment (michaelforney) [inline build.c]: Let's call this `starttime`.

File: build.c [hunk 5]
```diff
@@ -199,6 +202,46 @@ buildadd(struct node *n)
...
+				printf("%zu", nstarted - nfinished);
+				break;
+			case 'u':
+				printf("%zu", ntotal - nstarted);
+				break;
+			case 'p':
+				printf("%3lu%%", 100 * nfinished / ntotal);
+				break;
+			case 'e':
+				clock_gettime(CLOCK_REALTIME, &ctime);
```
Comment (michaelforney) [inline build.c]: I think `CLOCK_MONOTONIC` is the right clock to use here. Then it won't be affected by a system clock change during the job (unlikely, but still).

**Comment by E5ten** [comment]
for the default NINJA_STATUS I went with "[%f/%t] " like ninja, but if you want I can change that to "[%s/%t] " to be in line with what samurai does now, it'll also

File: build.c [hunk 6]
```diff
@@ -199,6 +202,46 @@ buildadd(struct node *n)
...
+				printf("%.3f",
+					((double)ctime.tv_sec + (double)ctime.tv_nsec * 0.000000001) -
```
Comment (michaelforney) [inline build.c]: I don't think these casts are necessary. The `0.000000001` literal has type `double`, and usual arithmetic conversions will convert the integers to floating point for the binary operators.

**Comment by michaelforney** [comment]
Also, could you add a short description of `NINJA_STATUS` to the man page?

**Comment by michaelforney** [comment]
Pushed with a few more tweaks. Thanks again!

**Pull request closed without merging**